packages/myst-cli/src/build/utils/collectExportOptions.ts

```diff
@@ -1,69 +1,307 @@
 
-function assignArticlesFromProject(
-  exp: Export,
+type ResolvedArticles = Pick<Export, 'articles' | 'sub_articles'>;
+
+function resolveArticlesFromProject(
+  exp: ExportWithFormat,
   proj: Omit<LocalProject, 'bibliography'>,
   vfile: VFile,
-) {
+): ResolvedArticles {
   const { file, pages } = proj;
   const fileAsPage = { file, level: 1 };
   const articles = pages?.length ? pages : [fileAsPage];
   if (MULTI_ARTICLE_EXPORT_FORMATS.includes(exp.format)) {
-    exp.articles = articles;
-  } else if (exp.format === ExportFormats.xml) {
-    exp.articles = [fileAsPage];
-    exp.sub_articles = pages
-      .map((page) => (page as any).file)
-      .filter((pageFile): pageFile is string => !!pageFile);
+    return { articles };
+  }
+  if (exp.format === ExportFormats.xml) {
+    return {
+      articles: [fileAsPage],
+      sub_articles: pages
+        .map((page) => (page as any).file)
+        .filter((pageFile): pageFile is string => !!pageFile),
+    };
   } else {
     fileError(vfile, "multiple articles are only supported for 'tex', 'typst', and 'pdf' exports", {
       ruleId: RuleId.validFrontmatterExportList,
     });
-    exp.articles = [singleArticleWithFile(articles) ?? fileAsPage];
+    return { articles: [singleArticleWithFile(articles) ?? fileAsPage] };
   }
 }
 
-function assignArticlesFromTOC(session: ISession, exp: Export, tocPath: string, vfile: VFile) {
+function resolveArticlesFromTOC(
+  session: ISession,
+  exp: ExportWithFormat,
+  tocPath: string,
+  vfile: VFile,
+): ResolvedArticles {
   const allowLevelLessThanOne = [
     ExportFormats.tex,
     ExportFormats.pdf,
     ExportFormats.pdftex,
   ].includes(exp.format);
   const proj = projectFromToc(session, tocPath, allowLevelLessThanOne ? -1 : 1);
-  assignArticlesFromProject(exp, proj, vfile);
+  return resolveArticlesFromProject(exp, proj, vfile);
+}
+
+/**
+ * Resolve template to absolute path
+ *
+ * If `disableTemplate: true` template resolves to `null`
+ *
+ * Otherwise, if it exists on the path relative to the source file, it is
+ * resolved to absolute path. If not, it is unchanged.
+ */
+function resolveTemplate(
+  sourceFile: string,
+  exp: Export,
+  disableTemplate?: boolean,
+): string | null | undefined {
+  if (disableTemplate) return null;
+  if (exp.template) {
+    const resolvedTemplatePath = path.resolve(path.dirname(sourceFile), exp.template);
+    if (fs.existsSync(resolvedTemplatePath)) {
+      return resolvedTemplatePath;
+    }
+  }
+  return exp.template;
+}
+
+/**
+ * Resolve format based on export format/template/output
+ *
+ * User-provided `format` is prioritized, then `template` type, then `output` file extension.
+ * (Validation ensures one of format/output/template is provided.)
+ *
+ * If `format` or `output` give PDF export, we look at the template kind and switch to `typst`
+ * in the case of a typst template.
+ */
+function resolveFormat(vfile: VFile, exp: Export): ExportFormats | undefined {
```

Comment:
I need to add some unit tests at least on these `resolve*` functions before merging this...